### Rad sa duplikatima

In [2]:
import pandas as pd

#### Koristimo skup podataka o avionskim nesrećama i posmatramo kolonu datum.

In [4]:
df = pd.read_csv('airAccs.csv')

In [5]:
df.columns

Index(['Date', 'location', 'operator', 'planeType', 'Dead', 'Aboard',
       'Ground'],
      dtype='object')

In [6]:
df.head()

Date                            location                operator  \
0   9/17/1908                 Fort Myer, Virginia    Military - U.S. Army   
1   7/12/1912           Atlantic City, New Jersey    Military - U.S. Navy   
2    8/6/1913  Victoria, British Columbia, Canada                 Private   
3    9/9/1913                  Over the North Sea  Military - German Navy   
4  10/17/1913          Near Johannisthal, Germany  Military - German Navy   

                planeType  Dead  Aboard  Ground  
0        Wright Flyer III   1.0     2.0     0.0  
1               Dirigible   5.0     5.0     0.0  
2        Curtiss seaplane   1.0     1.0     0.0  
3  Zeppelin L-1 (airship)  14.0    20.0     0.0  
4  Zeppelin L-2 (airship)  30.0    30.0     0.0

#### Pomoću metoda nunique proverićemo da li imamo duplikate za kolonu datum.

In [8]:
df.nunique()

Date         5100
location     4622
operator     2729
planeType    2663
Dead          194
Aboard        238
Ground         53
dtype: int64

In [9]:
df.shape

(5666, 7)

#### Vidimo da imamo 5666 redova u skup podataka, ali samo 5100 jedinstevnih datuma. To znači da se neki datumi ponavljaju, tj. da imamo duplikate u skupu podataka.

#### Osim ako nisu očigledni duplikati (redovi koji imaju iste vrednosti za sve atribute), potrebno je pažljivo proučiti svaki potencijalni duplikat pre nego što se eventualno odlučimo da ga uklonimo.

#### Koristimo metod duplicated kome prosleđujemo jedan ili više atributa, a kao rezultat dobijamo True za svaki red koji je duplikat za taj atribut ili atribute. Sa parametrom keep određujemo da li nam u rezultatu duplicated ostaje jedno (opcije keep=First ili keep=Last) ili sva ponavljanja reda koji je duplikat.

In [11]:
duplicate_rows = df[df.duplicated(subset=['Date'], keep=False)]

duplicate_rows

Date                        location  \
35     8/16/1920          College Park, Maryland   
36     8/16/1920                Bedford, England   
119    4/22/1927                   Floh, Germany   
120    4/22/1927                 Goshen, Indiana   
125    9/17/1927              Hadley, New Jersey   
...          ...                             ...   
5623  12/25/2012                   Heho, Myanmar   
5624  12/25/2012       Near Shymkent, Kazakhstan   
5651  11/29/2013                Glasgow Scotland   
5652  11/29/2013  Bwabwata National Park, Nambia   
5653  11/29/2013          Near St. Marys, Alaska   

                                 operator                  planeType  Dead  \
35                 US Aerial Mail Service          De Havilland DH-4   1.0   
36                                 By Air  Armstrong-Whitworth F-K-8   1.0   
119                    Deutsche Lufthansa              Fokker FG III   2.0   
120                US Aerial Mail Service                Douglas M-4   1.0   
125                      Reynolds Airways               Fokker F-VII   7.0   
...                                   ...                        ...   ...   
5623                            Air Bagan                 Fokker 100   2.0   
5624  Military - Kazakhstan Border Guards         Antonov AN-72-100D  27.0   
5651               Bond Air Services Ltd.        Eurocopter EC135 T2   3.0   
5652                   Mozambique Airline          Embraer ERJ-190AR  33.0   
5653           Hageland Aviation Services  Cessna 208B Grand Caravan   4.0   

      Aboard  Ground  
35       1.0     0.0  
36       1.0     0.0  
119      2.0     0.0  
120      1.0     0.0  
125     12.0     0.0  
...      ...     ...  
5623    71.0     1.0  
5624    27.0     0.0  
5651     3.0     5.0  
5652    33.0     0.0  
5653    10.0     0.0  

[1080 rows x 7 columns]

In [12]:
duplicate_rows = df[df.duplicated(subset=['Date', 'operator'], keep=False)]

duplicate_rows

Date                               location  \
125    9/17/1927                     Hadley, New Jersey   
126    9/17/1927              Near Dunellen, New Jersey   
127    9/23/1927                       Schleiz, Germany   
128    9/23/1927      Near schleiz, Thuringia,, Germany   
139    1/22/1928                                 France   
140    1/22/1928                       Tarragona, Spain   
301    1/20/1933                          Bourne, Texas   
302    1/20/1933                      Marietta, Georgia   
511    10/1/1938                   Off Bathurst, Gambia   
512    10/1/1938                   Grisons, Switzerland   
523    12/8/1938                         East China Sea   
524    12/8/1938                   Off Kubashima, Japan   
671   11/19/1943                         Kunming, China   
672   11/19/1943                         Kunming, China   
690     6/6/1944                       Carentan, France   
691     6/6/1944                    Flamanville, France   
692     6/6/1944                     Barneville, France   
886   12/25/1946                        Shanghai, China   
888   12/25/1946                   Near Shanghai, China   
1669    9/2/1958                               Off Guam   
1671    9/2/1958                      Sasnashen, Russia   
1719   5/12/1959              Charleston, West Virginia   
1720   5/12/1959                        Chase, Maryland   
1769   2/26/1960                  Lvov-Snilow,  Ukraine   
1770   2/26/1960                    Lvov, Ukraine, USSR   
1800   8/17/1960                    Kiev, Ukraine, USSR   
1802   8/17/1960                     Near Kiev, Ukraine   
1898   3/15/1962  Western Pacific Ocean, Philippine Sea   
1899   3/15/1962                           Adak, Alaska   
2183   2/17/1966                         Moscow, Russia   
2184   2/17/1966                         Moscow, Russia   
2357   2/29/1968                    Near Bratsk, Russia   
2358   2/29/1968                    Near Bratsk, Russia   
2592    9/2/1970                       Lihovka, Ukraine   
2593    9/2/1970     Near Dnepropetrovsk, Ukraine, USSR   
2811   2/28/1973                       Szczecin, Poland   
2812   2/28/1973                  Near Szczecin, Poland   
3104   8/28/1976            Sondre Stromfjord,Greenland   
3106   8/28/1976                  Near Thorney, England   
3132  12/17/1976                        Ust-Kut, Russia   
3133  12/17/1976                             Kiev, USSR   
4079   9/21/1989      Centeral Afghanistan\nAfghanistan   
4080   9/21/1989              Near Baghran, Afghanistan   
5003   9/11/2001                New York City, New York   
5004   9/11/2001                New York City, New York   
5005   9/11/2001              Shanksville, Pennsylvania   
5006   9/11/2001                   Arlington, Virginia.   

                                    operator  \
125                         Reynolds Airways   
126                         Reynolds Airways   
127                       Deutsche Lufthansa   
128                       Deutsche Lufthansa   
139                              Aeropostale   
140                              Aeropostale   
301                         American Airways   
302                         American Airways   
511                       Deutsche Lufthansa   
512                       Deutsche Lufthansa   
523               Japan Aviation Corporation   
524               Japan Aviation Corporation   
671      China National Aviation Corporation   
672      China National Aviation Corporation   
690          Military - U.S. Army Air Forces   
691          Military - U.S. Army Air Forces   
692          Military - U.S. Army Air Forces   
886                                     CNAC   
888                                     CNAC   
1669               Military - U.S. Air Force   
1671               Military - U.S. Air Force   
1719                        Capital Airlines   
1720                        Capital Airlines   
1769                       

In [13]:
len(duplicate_rows)

47

#### Rezultate smo suzili na 47 koje bi trebalo ručno obraditi

#### Na primer, redovi 125 i 126 imaju isti datum i državu, dok su gradovi veoma blizu jedan drugog (15 minuta vožnje po Google maps). Za ovaj primer možemo sa velikom sigurnošću da tvrdimo da je duplikat.
#### U ovom primeru pored odvajanja gradova i države bilo bi korisno dodati atribute o geografskoj lokaciji grada kako bi mogli da još više automatizujemo određivanja udaljenosti gradova. Ako se ne može pronaći gotov skup podataka sa geo. lokacijama, sigurno se može prikupiti pomoću web scraping alata. Takav primer ćemo imati na sledećem predavanju.

#### Dodavanje geo lokacija i izdvajanje države su upravu lepi primeri data wrangling procesa.

#### Za kraj rada sa ovim skupom podataka testiramo da li postoje očigledni duplikati.

In [15]:
duplicate_all_df = df.loc[df.duplicated(keep=False).values] 
duplicate_all_df
#df.duplicated(keep=False).values imaće vrednosti True samo za redove koji su duplikati po svim atributima

Empty DataFrame
Columns: [Date, location, operator, planeType, Dead, Aboard, Ground]
Index: []

#### U prethodnom primeru videli smo da nema redova koji su duplikati po svim atributima. Pogledaćemo sada još jedan primer. U pitanju je skup podataka o pacijentima koji imaju rak dojke.

In [17]:
df_bc = pd.read_csv('breast_cancer_data.csv')

In [18]:
df_bc.head(5)

patient_id  clump_thickness  cell_size_uniformity  cell_shape_uniformity  \
0     1000025              5.0                   1.0                      1   
1     1002945              5.0                   4.0                      4   
2     1015425              3.0                   1.0                      1   
3     1016277              6.0                   8.0                      8   
4     1017023              4.0                   1.0                      1   

   marginal_adhesion  single_ep_cell_size bare_nuclei  bland_chromatin  \
0                  1                    2           1              3.0   
1                  5                    7          10              3.0   
2                  1                    2           2              3.0   
3                  1                    3           4              3.0   
4                  3                    2           1              3.0   

   normal_nucleoli  mitoses   class doctor_name  
0              1.0        1  benign     Dr. Doe  
1              2.0        1  benign   Dr. Smith  
2              1.0        1  benign     Dr. Lee  
3              7.0        1  benign   Dr. Smith  
4              1.0        1  benign    Dr. Wong

In [19]:
df_bc.nunique()

patient_id               645
clump_thickness           10
cell_size_uniformity      10
cell_shape_uniformity     10
marginal_adhesion         10
single_ep_cell_size       10
bare_nuclei               11
bland_chromatin           10
normal_nucleoli           10
mitoses                    9
class                      2
doctor_name                4
dtype: int64

In [20]:
df_bc.shape

(699, 12)

#### Vidimo da od imamo 699 redova u skupu podataka, ali samo 645 jedinstvenih identifikatora pacijenata (patient_id). Trebalo bi da proverimo da li su u pitanju stvarno duplikati ili je možda došlo do greške prilikom unosa identifikatora za neke od pacijenata. 

In [22]:
duplicate_patients = df_bc[df_bc.duplicated(subset=['patient_id'], keep=False).values].sort_values('patient_id')
duplicate_patients.head(10)

patient_id  clump_thickness  cell_size_uniformity  cell_shape_uniformity  \
267      320675              3.0                   3.0                      5   
272      320675              3.0                   3.0                      5   
575      385103              5.0                   1.0                      2   
269      385103              1.0                   1.0                      1   
271      411453              5.0                   1.0                      1   
607      411453              1.0                   1.0                      1   
684      466906              1.0                   1.0                      1   
683      466906              1.0                   1.0                      1   
371      493452              1.0                   1.0                      3   
372      493452              4.0                   1.0                      2   

     marginal_adhesion  single_ep_cell_size bare_nuclei  bland_chromatin  \
267                  2                    3          10              7.0   
272                  2                    3          10              7.0   
575                  1                    2           1              3.0   
269                  1                    2           1              3.0   
271                  1                    2           1              3.0   
607                  1                    2           1              1.0   
684                  1                    2           1              1.0   
683                  1                    2           1              1.0   
371                  1                    2           1              1.0   
372                  1                    2           1              2.0   

     normal_nucleoli  mitoses      class doctor_name  
267              1.0        1  malignant    Dr. Wong  
272              1.0        1  malignant   Dr. Smith  
575              1.0        1     benign   Dr. Smith  
269              1.0        1     benign     Dr. Doe  
271              1.0        1     benign    Dr. Wong  
607              1.0        1     benign   Dr. Smith  
684              1.0        1     benign    Dr. Wong  
683              1.0        1     benign     Dr. Lee  
371              1.0        1     benign   Dr. Smith  
372              1.0        1     benign    Dr. Wong

#### Pre nego što protumačimo duplikate po atributu id, proverićemo sada da li postoje duplikati po svim atributima.

In [24]:
duplicate_all_df_bc = df_bc.loc[df_bc.duplicated(keep=False).values]
duplicate_all_df_bc

patient_id  clump_thickness  cell_size_uniformity  cell_shape_uniformity  \
168     1198641              3.0                   1.0                      1   
258     1198641              3.0                   1.0                      1   

     marginal_adhesion  single_ep_cell_size bare_nuclei  bland_chromatin  \
168                  1                    2           1              3.0   
258                  1                    2           1              3.0   

     normal_nucleoli  mitoses   class doctor_name  
168              1.0        1  benign     Dr. Lee  
258              1.0        1  benign     Dr. Lee

#### Vidmo da imamo jedan duplikat kod koga su svi atributi jednaki, dok se kod ostalih redovi razlikuju po imenu lekara i/ili nekim od vrednosti dijagonostičkih testova. 
#### Možemo na primer da pretpostavimo da su ovi redovi korektni i da je pacijent sa istim identifikatorom više puta dolazio u bolincu gde je njegovo stanje praćeno - što se može videti iz tabele ispod gde se vidi da su promene u atributima kao što su bland_chromatin ili clump_thicknes itd. 
#### Naravno bez uvida u meta-podatke skupa podataka ili razgovora sa kretorima skupa podataka ne možemo sa sigurnošću da znamo koje duplikate treba ukloniti, a koje ne. 

In [69]:
duplicate_rows = df_bc[df_bc.duplicated(subset=['patient_id', 'doctor_name', 'class'], keep=False)].sort_values('patient_id', ascending=False)

duplicate_rows

patient_id  clump_thickness  cell_size_uniformity  cell_shape_uniformity  \
516     1320077              1.0                   1.0                      1   
517     1320077              1.0                   1.0                      1   
431     1276091              5.0                   1.0                      1   
430     1276091              1.0                   3.0                      1   
241     1276091              3.0                   1.0                      1   
548     1240603              3.0                   1.0                      1   
547     1240603              2.0                   1.0                      1   
471     1238777              6.0                   1.0                      1   
632     1238777              1.0                   1.0                      1   
258     1198641              3.0                   1.0                      1   
168     1198641              3.0                   1.0                      1   
256     1182404              3.0                   1.0                      1   
136     1182404              4.0                   1.0                      1   
448     1182404              1.0                   1.0                      1   
497     1182404              4.0                   2.0                      1   
265     1182404              5.0                   1.0                      4   
255     1168736              5.0                   6.0                      6   
104     1168736             10.0                  10.0                     10   
82      1143978              5.0                   2.0                      1   
81      1143978              4.0                   1.0                      1   
387     1114570              5.0                   3.0                      3   
388     1114570              2.0                   1.0                      1   
47      1105524              1.0                   1.0                      1   
468     1105524              4.0                   1.0                      1   
4       1017023              4.0                   1.0                      1   
252     1017023              6.0                   3.0                      3   
290      560680              1.0                   1.0                      1   
374      560680              3.0                   1.0                      2   

     marginal_adhesion  single_ep_cell_size bare_nuclei  bland_chromatin  \
516                  1                    1           1              1.0   
517                  1                    1           1              2.0   
431                  3                    4           1              3.0   
430                  1                    2           1              2.0   
241                  3                    1           1              3.0   
548                  1                    1           1              1.0   
547                  1                    1           1              1.0   
471                  3                    2           1              1.0   
632                  1                    2           1              1.0   
258                  1                    2           1              3.0   
168                  1                    2           1              3.0   
256                  1                    2           1              1.0   
136                  1                    2           1              2.0   
448                  1                    1           1              1.0   
497                  1                    2           1              1.0   
265                  1                    2           1              3.0   
255                  2                    4          10              3.0   
104                 10                   10           1              8.0   
82                   1                    2           1              3.0   
81                   2                    2           1              2.0   
387                  2                    3           1    

In [65]:
df_bc[df_bc.duplicated(subset=['patient_id', 'class'], keep=False)].groupby('patient_id').size().sort_values(ascending=False)

patient_id
1182404    6
1276091    5
320675     2
1218860    2
1143978    2
1158247    2
1168736    2
1173347    2
1174057    2
1198641    2
1212422    2
1238777    2
1116116    2
1240603    2
1277792    2
1293439    2
1299924    2
1320077    2
1321942    2
1339781    2
1116192    2
1115293    2
385103     2
769612     2
411453     2
466906     2
493452     2
560680     2
654546     2
704097     2
733639     2
734111     2
798429     2
1114570    2
822829     2
897471     2
1017023    2
1033078    2
1061990    2
1070935    2
1100524    2
1105524    2
1354840    2
dtype: int64